### PR TITLE
Wrap binstats

### PR DESCRIPTION
**Pull request by willschlitzer**
This wraps the GMT module `gmtbinstats`.

<!-- If fixing an issue, put the issue number after the # below (no spaces). GitHub will automatically close it when this gets Addresses #1651

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

File: pygmt/src/binstats.py
Comment (seisman) [inline pygmt/src/binstats.py:98]:
```suggestion
                if (outgrid := kwargs.get("G")) is None:
                    kwargs["G"] = outgrid = tmpfile.name  # output to tmpfile
```

File: pygmt/src/binstats.py
Comment (seisman) [inline pygmt/src/binstats.py:100]:
```suggestion
                lib.call_module("gmtbinstats", build_arg_string(kwargs, infile=infile))
```

File: pygmt/src/binstats.py
Comment (seisman) [inline pygmt/src/binstats.py:98]:
```suggestion
                lib.call_module(
                    module="binstats", 
                    args=build_arg_string(kwargs, infile=infile)
                )
```

File: pygmt/src/binstats.py
Comment (seisman) [inline pygmt/src/binstats.py:39]:
```suggestion
    (2-4 columns) from standard input [or ``data``] and for each
```

File: pygmt/src/binstats.py
Comment (seisman) [inline pygmt/src/binstats.py:54]: >         Pass in data with L, M, C[L,M], S[L,M] values by

What does this mean?

File: pygmt/src/binstats.py
Comment (seisman) [inline pygmt/src/binstats.py:17]: I feel that `statistic` is a better alias for `-C`.

File: pygmt/src/binstats.py
Comment (willschlitzer) [reply pygmt/src/binstats.py:54]: Not sure if I copied the text from the incorrect GMT file.  I'll remove it.

File: pygmt/src/binstats.py
Comment (weiji14) [inline pygmt/src/binstats.py:26]: Still missing some parameter docstrings, need to copy from https://github.com/GenericMappingTools/gmt/blob/6.3.0/doc/rst/source/gmtbinstats.rst.

- [ ] normalize (N)
- [ ] search_radius (S)
- [ ] tiling (T)
- [ ] weight (W)

File: pygmt/src/binstats.py
Comment (weiji14) [inline pygmt/src/binstats.py:39]:
```suggestion
    (2-4 columns) from ``data``] and for each
```

File: pygmt/src/binstats.py
Comment (seisman) [inline pygmt/src/binstats.py:39]:
```suggestion
    (2-4 columns) from ``data`` and for each
```

File: pygmt/src/binstats.py
Comment (seisman) [inline pygmt/src/binstats.py:45]: I think this is a typo in the GMT documentation https://docs.generic-mapping-tools.org/dev/gmtbinstats.html:
```suggestion
    hexagonal tiling instead of a rectangular grid layout.
```

File: pygmt/src/binstats.py
Comment (seisman) [inline pygmt/src/binstats.py:101]: `{x}` is not available in gmtbinstats.
```suggestion
```

File: pygmt/src/binstats.py
Comment (seisman) [inline pygmt/src/binstats.py:18]: Still missing the docstring for `empty`.

File: pygmt/src/binstats.py
Comment (seisman) [inline pygmt/src/binstats.py:89]: > For **h**, a table with the centers of the hexagons and the computed statistics is written to the ouput.

By default, `binstats` generates a grid, but when `tiling="h"`, it outputs some information to an ASCII file instead. Because it behaves differently, I think it would be better NOT to add the alias `tiling` in this PR.

File: pygmt/src/binstats.py
Comment (seisman) [inline pygmt/src/binstats.py:94]:
```suggestion
        number of points. If the weights are actually uncertainties
```
```suggestion
        number of points. If tje weights are actually uncertainties
```

Comment (seisman) [inline pygmt/src/binstats.py:94]:
```suggestion
        number of points. If the weights are actually uncertainties
```
```suggestion
        number of points. If tje weights are actually uncertainties
```

File: pygmt/src/binstats.py
Comment (seisman) [inline pygmt/src/binstats.py:24]: Remove alias `tiling` from this PR.
```suggestion
```

File: pygmt/src/binstats.py
Comment (seisman) [inline pygmt/src/binstats.py:45]:
```suggestion
    perhaps give density estimates.
```

File: pygmt/src/binstats.py
Comment (seisman) [inline pygmt/src/binstats.py:89]:
```suggestion
```

File: pygmt/src/binstats.py
Comment (seisman) [inline pygmt/src/binstats.py:26]:
```suggestion
    a="aspatial",
```

File: pygmt/tests/test_binstats.py
Comment (seisman) [inline pygmt/tests/test_binstats.py:29]:
```suggestion
def test_binstats_no_outgrid():
```

**Review by seisman — APPROVED** [review]

**Review by seisman — APPROVED** [review]
Looks good to me after fixing a typo.

**Comment by seisman** [comment]
@GenericMappingTools/pygmt-maintainers Hopefully someone else can give this PR a final review.

File: pygmt/tests/test_binstats.py
Comment (meghanrjones) [inline pygmt/tests/test_binstats.py:34]: Can you add this to pygmt.helpers.testing.download_test_data?

**Review by meghanrjones — APPROVED** [review]
Great work, @willschlitzer!

The only important comment here is to include `@capitals.gmt` in the `download_test_data` function since that's used for caching the files for the CI.

I agree with https://github.com/GenericMappingTools/pygmt/pull/1652#discussion_r881805969 that the output to file option can be added later since it's more similar to https://github.com/GenericMappingTools/pygmt/issues/896 than https://github.com/GenericMappingTools/pygmt/issues/1536.

There's some work that needs to be done for managing required arguments, but I think we need some more consistency regarding that throughout the library and it's not a reason to hold this feature back.

**Pull request merged**
